File: src/main/intellibitz/sted/ui/STEDWindow.java
package sted.ui;

import sted.actions.ExitAction;
import sted.actions.ItemListenerAction;
import sted.actions.STEDWindowAction;
import sted.event.*;
import sted.fontmap.FontMap;
import sted.io.FileReaderThread;
import sted.util.FileHelper;
import sted.util.MenuHandler;
import sted.util.Resources;

import javax.swing.*;
import javax.swing.event.ChangeEvent;
import javax.swing.event.ChangeListener;
import java.awt.*;
import java.util.ArrayList;
import java.util.Map;
import java.util.logging.Logger;

public class STEDWindow
        extends JFrame
        implements IThreadListener,
        ChangeListener,
        IMessageListener,
        IStatusEventSource {
    // desktop to show fontmap
    private TabDesktop tabDesktop;
    private StatusPanel statusPanel;
    private static final Logger logger =
            Logger.getLogger(STEDWindow.class.getName());
    private IStatusListener statusListener;
    private StatusEvent statusEvent;

    public STEDWindow() {
        super();
    }

    public void init() {
        setTitle(Resources.getSTEDTitle());
        setDefaultLookAndFeelDecorated(true);
        setState(JFrame.MAXIMIZED_BOTH);
        setExtendedState(JFrame.MAXIMIZED_BOTH);
        setDefaultCloseOperation(JFrame.DO_NOTHING_ON_CLOSE);
        setIconImage(Resources.getSTEDImage());

        statusEvent = new StatusEvent(this);

        final JMenuBar menuBar =
                MenuHandler.getInstance().getMenuBar(Resources.MENUBAR_STED);
        MenuHandler.loadLookAndFeelMenu(this);

        // load the menubar for the application
        setJMenuBar(menuBar);
        fireStatusPosted("20");

        final Container container = getContentPane();
        final GridBagLayout gridBagLayout = new GridBagLayout();
        container.setLayout(gridBagLayout);
        final GridBagConstraints gridBagConstraints = new GridBagConstraints();
        gridBagConstraints.gridwidth = 1;
        gridBagConstraints.gridheight = 1;
        gridBagConstraints.weightx = 1.0;
        gridBagConstraints.weighty = 0.0;
        gridBagConstraints.gridx = 0;
        gridBagConstraints.gridy = 0;
        gridBagConstraints.fill = GridBagConstraints.HORIZONTAL;
        fireStatusPosted("30");

        final JToolBar toolBar =
                MenuHandler.getInstance().getToolBar(Resources.MENUBAR_STED);
        gridBagLayout.setConstraints(toolBar, gridBagConstraints);

        // adds the toolbar for the app
        container.add(toolBar);
        fireStatusPosted("40");

        tabDesktop = new TabDesktop();
        tabDesktop.init();
        fireStatusPosted("50");
        gridBagConstraints.weighty = 1.0;
        gridBagConstraints.gridy = 1;
        gridBagConstraints.fill = GridBagConstraints.BOTH;
        // adds the desktop directly
        gridBagLayout.setConstraints(tabDesktop, gridBagConstraints);

        container.add(tabDesktop);
        fireStatusPosted("60");

        statusPanel = new StatusPanel(this);
        gridBagConstraints.weighty = 0.0;
        gridBagConstraints.gridy = 2;
        gridBagConstraints.fill = GridBagConstraints.BOTH;
        gridBagLayout.setConstraints(statusPanel, gridBagConstraints);

        // adds the status bar
        container.add(statusPanel);
        fireStatusPosted("70");
        setUserOptions();
        addMouseListener(AboutSTED.getInstance());

        ExitAction exitAction = new ExitAction();
        exitAction.setSTEDWindow(STEDWindow.this);
        addWindowListener(exitAction);

        pack();
        logger.finest("successfully intialized STEDWindow");
        fireStatusPosted("80");
    }

    public void load() {
        // status panel added as status listener to recieve status messages
        tabDesktop.addStatusListener(statusPanel);

        Map<String, Action> actions = MenuHandler.getInstance().getActions();
        for (Action action : actions.values()) {
            if (STEDWindowAction.class.isInstance(action)) {
                ((STEDWindowAction) action).addStatusListener(statusPanel);
                ((STEDWindowAction) action).addMessageListener(this);
            }
        }

        tabDesktop.load();
        tabDesktop.addChangeListener(this);

        setState(JFrame.MAXIMIZED_HORIZ);
        setExtendedState(JFrame.MAXIMIZED_BOTH);
    }

    public void fireStatusPosted(String message) {
        statusEvent.setStatus(message);
        statusListener.statusPosted(statusEvent);
    }

    public void fireStatusPosted() {
        statusListener.statusPosted(statusEvent);
    }

    public void addStatusListener(IStatusListener statusListener) {
        this.statusListener = statusListener;
    }

    public void setVisible() {
        super.setVisible(true);
        statusPanel.runMemoryBar();
    }

    public StatusPanel getStatusPanel() {
        return statusPanel;
    }

    public TabDesktop getDesktop() {
        return tabDesktop;
    }

    private void setUserOptions() {
        final Map<String, JMenuItem> menuItems =
                MenuHandler.getInstance().getMenuItems();
        for (String key : menuItems.keySet()) {
            final JMenuItem menuItem = menuItems.get(key);
            final Action action = menuItem.getAction();
            if (ItemListenerAction.class.isInstance(action)) {
                final String val = Resources.getSetting(key);
                if (val != null) {
                    final boolean curr = Boolean.valueOf(val);
                    // first check for state for disabled menu items
                    if (!action.isEnabled()) {
                        menuItem.setSelected(curr);
                    } else if (curr && !menuItem.isSelected()) {
                        // if user option is set, and the menu item is already not selected
                        menuItem.doClick();
                    } else if (menuItem.isSelected() && !curr) {
                        menuItem.doClick();
                    }
                }
            }
        }
        final ArrayList<String> reopenItems = Resources
                .getSettingBeginsWith(Resources.ACTION_FILE_REOPEN_COMMAND);
        if (!reopenItems.isEmpty()) {
            final JMenu menu =
                    MenuHandler.getInstance()
                            .getMenu(Resources.ACTION_FILE_REOPEN_COMMAND);
            for (String reopenItem : reopenItems) {
                MenuHandler.addReOpenItem(menu, reopenItem);
            }
            menu.setEnabled(menu.getItemCount() > Resources.DEFAULT_MENU_COUNT);
        }

        // set the sample fontmap action
        String[] sampleFontMapPaths = FileHelper
                .getSampleFontMapPaths(Resources.getResourceDirPath());
        if (sampleFontMapPaths.length > 0) {
            final JMenu menu =
                    MenuHandler.getInstance()
                            .getMenu(Resources.MENU_SAMPLES_NAME);
            for (String reopenItem : sampleFontMapPaths) {
                MenuHandler.addSampleFontMapMenuItem(menu, reopenItem);
            }
        }
    }

    public void threadRunStarted(ThreadEvent e) {
        JProgressBar progressBar = statusPanel.getProgressBar();
        progressBar.setMinimum(0);
        progressBar.setIndeterminate(true);
    }

    public void threadRunning(ThreadEvent e) {
        //TODO: set the progress bar maximum only once, typically when the thread starts
        JProgressBar progressBar = statusPanel.getProgressBar();
        progressBar.setMaximum(e.getEventSource().getProgressMaximum());
        progressBar.setValue(e.getEventSource().getProgress());
    }

    public void threadRunFailed(ThreadEvent e) {
        JOptionPane.showMessageDialog(this, e.getEventSource().getMessage());
        JProgressBar progressBar = statusPanel.getProgressBar();
        progressBar.setValue(0);
        progressBar.setIndeterminate(false);
    }

    public void threadRunFinished(ThreadEvent e) {
        JProgressBar progressBar = statusPanel.getProgressBar();
        progressBar.setValue(0);
        progressBar.setIndeterminate(false);
        FileReaderThread source = (FileReaderThread) e.getEventSource();
        statusPanel.setStatus("Read File: " + source.getFile());
    }

    /**
     * @param e ChangeEvent published by TabDesktop
     */
    public void stateChanged(ChangeEvent e) {
        TabDesktop desktop = (TabDesktop) e.getSource();
        int index = desktop.getSelectedIndex();
        if (index > -1) {
            DesktopFrame dframe =
                    (DesktopFrame) desktop.getComponentAt(
                            index);
            dframe.getInputFileViewer().addThreadListener(this);
            dframe.getOutputFileViewer().addThreadListener(this);

            // update the lock icon
            DesktopModel desktopModel =
                    dframe.getModel();
            final FontMap fontMap = desktopModel.getFontMap();

            fontMap.removeFontMapChangeListener(statusPanel);
            fontMap.addFontMapChangeListener(statusPanel);

            statusPanel.setLockFlag(!fontMap.isFileWritable());
            // update the clean/dirty flag
            statusPanel.setNeatness(fontMap);
            dframe.getMapperPanel().getMappingEntryPanel().getEntryAction()
                    .addStatusListener(statusPanel);

            dframe.getMapperPanel().getMappingEntryPanel().getEntryAction()
                    .addMessageListener(this);
            dframe.getMapperPanel().getMappingEntryPanel()
                    .getMappingTableModel().addMessageListener(this);

            // remove and add.. so getting added only once
            dframe.getMapperPanel().getMappingEntryPanel().
                    getMappingTableModel()
                    .removeTableModelListener(statusPanel);
            dframe.getMapperPanel().getMappingEntryPanel().
                    getListSelectionModel()
                    .removeListSelectionListener(statusPanel);
            dframe.getMapperPanel().getMappingEntryPanel().
                    getMappingTableModel().addTableModelListener(statusPanel);
            dframe.getMapperPanel().getMappingEntryPanel().
                    getListSelectionModel()
                    .addListSelectionListener(statusPanel);
        }
    }

    public void messagePosted(MessageEvent event) {
        JOptionPane.showMessageDialog(this, event.getMessage());
    }
}

File: src/main/intellibitz/sted/ui/STEDWindow.kt
package sted.ui

import sted.actions.ExitAction
import sted.actions.ItemListenerAction
import sted.actions.STEDWindowAction
import sted.event.*
import sted.io.FileHelper.getSampleFontMapPaths
import sted.io.FileReaderThread
import sted.io.Resources
import sted.io.Resources.getResource
import sted.io.Resources.getSetting
import sted.io.Resources.getSettingBeginsWith
import sted.io.Resources.getSystemResourceIcon
import sted.ui.MenuHandler.Companion.addReOpenItem
import sted.ui.MenuHandler.Companion.addSampleFontMapMenuItem
import sted.ui.MenuHandler.Companion.loadLookAndFeelMenu
import java.awt.GridBagConstraints
import java.awt.GridBagLayout
import java.util.logging.LogManager
import java.util.logging.Logger
import javax.swing.JFrame
import javax.swing.JOptionPane
import javax.swing.event.ChangeEvent
import javax.swing.event.ChangeListener

class STEDWindow : JFrame(), IThreadListener, ChangeListener, IMessageListener, IStatusEventSource {
    // desktop to show fontmap
    val desktop: TabDesktop = TabDesktop()
    val statusPanel: StatusPanel = StatusPanel()
    private lateinit var statusEvent: StatusEvent
    private lateinit var statusListener: IStatusListener
    lateinit var logManager: LogManager

    val logger: Logger = Logger.getLogger(STEDWindow::class.java.name)
    fun init() {
        setDefaultLookAndFeelDecorated(true)
        state = MAXIMIZED_BOTH
        extendedState = MAXIMIZED_BOTH
        defaultCloseOperation = DO_NOTHING_ON_CLOSE
        title = getResource("title")
        val imageIcon = getSystemResourceIcon(getSetting("icon.sted"))
        if (imageIcon != null) iconImage = imageIcon.image
        statusEvent = StatusEvent(this)
        logManager.addLogger(logger)
        val xml = getResource("config.menu")
        if (xml == null) {
            logger.severe("Load menu file not found: please check config.menu property to set menu file")
        } else {
            MenuHandler.menuHandler.loadMenu(xml)
        }
        val menuBar = MenuHandler.menuHandler.getMenuBar("STED-MenuBar")
        loadLookAndFeelMenu()

        // load the menubar for the application
        jMenuBar = menuBar
        fireStatusPosted("20")
        val container = contentPane
        val gridBagLayout = GridBagLayout()
        container.layout = gridBagLayout
        val gridBagConstraints = GridBagConstraints()
        gridBagConstraints.gridwidth = 1
        gridBagConstraints.gridheight = 1
        gridBagConstraints.weightx = 1.0
        gridBagConstraints.weighty = 0.0
        gridBagConstraints.gridx = 0
        gridBagConstraints.gridy = 0
        gridBagConstraints.fill = GridBagConstraints.HORIZONTAL
        fireStatusPosted("30")
        val toolBar = MenuHandler.menuHandler.getToolBar(Resources.MENUBAR_STED)
        gridBagLayout.setConstraints(toolBar, gridBagConstraints)

        // adds the toolbar for the app
        container.add(toolBar)
        fireStatusPosted("40")

        desktop.init()
        fireStatusPosted("50")
        gridBagConstraints.weighty = 1.0
        gridBagConstraints.gridy = 1
        gridBagConstraints.fill = GridBagConstraints.BOTH
        // adds the desktop directly
        gridBagLayout.setConstraints(desktop, gridBagConstraints)
        container.add(desktop)
        fireStatusPosted("60")
        statusPanel.load(this)
        gridBagConstraints.weighty = 0.0
        gridBagConstraints.gridy = 2
        gridBagConstraints.fill = GridBagConstraints.BOTH
        gridBagLayout.setConstraints(statusPanel, gridBagConstraints)

        // adds the status bar
        container.add(statusPanel)
        fireStatusPosted("70")
        setUserOptions()
        addMouseListener(AboutSTED.instance)
        val exitAction = ExitAction()
        //        exitAction.setSTEDWindow(STEDWindow.this);
        addWindowListener(exitAction)
        pack()
        logger.finest("successfully intialized STEDWindow")
        fireStatusPosted("80")
    }

    fun load() {
        // status panel added as status listener to recieve status messages
        desktop.addStatusListener(statusPanel)
        val actions = MenuHandler.menuHandler.actions
        for (action in actions.values) {
            if (action is STEDWindowAction) {
                action.addStatusListener(statusPanel)
                action.addMessageListener(this)
            }
        }
        desktop.load()
        desktop.addChangeListener(this)
        state = MAXIMIZED_HORIZ
        extendedState = MAXIMIZED_BOTH
    }

    fun fireStatusPosted(message: String?) {
        statusEvent.status = message
        statusListener.statusPosted(statusEvent)
    }

    override fun fireStatusPosted() {
        statusListener.statusPosted(statusEvent)
    }

    override fun addStatusListener(statusListener: IStatusListener) {
        this.statusListener = statusListener
    }

    fun setVisible() {
        super.setVisible(true)
        statusPanel.runMemoryBar()
    }

    private fun setUserOptions() {
        val menuItems = MenuHandler.menuHandler.menuItems
        for (key in menuItems.keys) {
            val menuItem = menuItems[key]
            val action = menuItem!!.action
            if (action is ItemListenerAction) {
                val `val` = getSetting(key)
                if (`val` != null) {
                    val curr = `val`.toBoolean()
                    // first check for state for disabled menu items
                    if (!action.isEnabled()) {
                        menuItem.isSelected = curr
                    } else if (curr && !menuItem.isSelected) {
                        // if user option is set, and the menu item is already not selected
                        menuItem.doClick()
                    } else if (menuItem.isSelected && !curr) {
                        menuItem.doClick()
                    }
                }
            }
        }
        val reopenItems = getSettingBeginsWith(Resources.ACTION_FILE_REOPEN_COMMAND)
        if (reopenItems.isNotEmpty()) {
            val menu = MenuHandler.menuHandler.getMenu(Resources.ACTION_FILE_REOPEN_COMMAND)
            for (reopenItem in reopenItems) {
                addReOpenItem(menu!!, reopenItem)
            }
            menu!!.isEnabled = menu.itemCount > Resources.DEFAULT_MENU_COUNT
        }

        // set the sample fontmap action
        val sampleFontMapPaths = getSampleFontMapPaths("resource")
        if (sampleFontMapPaths?.size ?: 0 > 0) {
            val menu = MenuHandler.menuHandler.getMenu(Resources.MENU_SAMPLES_NAME)
            for (reopenItem in sampleFontMapPaths!!) {
                addSampleFontMapMenuItem(menu!!, reopenItem)
            }
        }
    }

    override fun threadRunStarted(threadEvent: ThreadEvent) {
        val progressBar = statusPanel.progressBar
        progressBar.minimum = 0
        progressBar.isIndeterminate = true
    }

    override fun threadRunning(threadEvent: ThreadEvent) {
        //TODO: set the progress bar maximum only once, typically when the thread starts
        val progressBar = statusPanel.progressBar
        progressBar.maximum = threadEvent.eventSource.progressMaximum
        progressBar.value = threadEvent.eventSource.progress
    }

    override fun threadRunFailed(threadEvent: ThreadEvent) {
        JOptionPane.showMessageDialog(this, threadEvent.eventSource.message)
        val progressBar = statusPanel.progressBar
        progressBar.value = 0
        progressBar.isIndeterminate = false
    }

    override fun threadRunFinished(threadEvent: ThreadEvent) {
        val progressBar = statusPanel.progressBar
        progressBar.value = 0
        progressBar.isIndeterminate = false
        val source = threadEvent.eventSource as FileReaderThread?
        statusPanel.setStatus("Read File: " + source!!.file)
    }

    /**
     * @param e ChangeEvent published by TabDesktop
     */
    override fun stateChanged(e: ChangeEvent) {
        val desktop = e.source as TabDesktop
        val index = desktop.selectedIndex
        if (index > -1) {
            val desktopFrame = desktop.getComponentAt(
                index
            ) as DesktopFrame
            desktopFrame.inputFileViewer.addThreadListener(this)
            desktopFrame.outputFileViewer.addThreadListener(this)

            // update the lock icon
            val desktopModel = desktopFrame.desktopModel
            val fontMap = desktopModel.fontMap
            fontMap.removeFontMapChangeListener(statusPanel)
            fontMap.addFontMapChangeListener(statusPanel)
            statusPanel.setLockFlag(!fontMap.isFileWritable)
            // update the clean/dirty flag
            statusPanel.setNeatness(fontMap)
            desktopFrame.mapperPanel.mappingEntryPanel.entryAction
                .addStatusListener(statusPanel)
            desktopFrame.mapperPanel.mappingEntryPanel.entryAction
                .addMessageListener(this)
            desktopFrame.mapperPanel.mappingEntryPanel
                .mappingTableModel.addMessageListener(this)

            // remove and add.. so getting added only once
            desktopFrame.mapperPanel.mappingEntryPanel.mappingTableModel
                .removeTableModelListener(statusPanel)
            desktopFrame.mapperPanel.mappingEntryPanel.listSelectionModel
                .removeListSelectionListener(statusPanel)
            desktopFrame.mapperPanel.mappingEntryPanel.mappingTableModel.addTableModelListener(statusPanel)
            desktopFrame.mapperPanel.mappingEntryPanel.listSelectionModel
                .addListSelectionListener(statusPanel)
        }
    }

    override fun messagePosted(messageEvent: MessageEvent) {
        JOptionPane.showMessageDialog(this, messageEvent.message)
    }

}